woocommerce/woocommerce-gateway-stripe
# Update Payment Request button after adding product to the cart

# Changes proposed in this Pull Request:

Part of #1563

- Changes the approach on how to respond to AJAX request for the Payment Request button. Now the information needed to display on the payment dialog (amount, shipping, taxes, display items, etc) will come from the backend asynchronously instead of relying of `wc_stripe_payment_request_params`. Unlike previous approach we will get that information from the cart now instead of calculating everything on the fly on the client or based on the product's price.
- `build_response()` will be in charge of building the data object used to respond to AJAX requests according to the state of the Payment Request button or what page the user is making the request.
- The behavior where product information was updated when changing the quantity input on the product page was removed. There's no reason to do that anymore because the real and latest payment information comes from the backend at the moment the user clicks the Payment Request button.

# Testing instructions

This PR focuses on simple products only. I've tried to test as many cases as I could but I might be missing something. Please make it break.

1. Create a simple product
2. Apply different scenarios to it. You can use the same cases I've tested on the Test list below.
3. Check correct amount and item list is correct in Google Pay/Apple Pay.

#### Test list - Simple Product

- [x] Regular Price
  - Note: When testing with no shipping zones (and without shipping methods) the "Shipping" display item is not displayed. It was displayed before but after picking an address the "Shipping" item was removed. I think the new behavior is better. No need to show the "Shipping" item if no shipping options are defined. Error is returned correctly when this happens
- [x] Errors correctly when no shipping zones are defined
- [x] Regular Price with quantity
- [x] With Sale price
- [x] With Sale price + Quantity
- [x] With shipping address
- [x] Virtual Product
- [x] Downloadable
  - Note: When updating the quantity the "Shipping" item is displayed correctly. Unlike the previous behavior the "Shipping" item was reset to `0` which was incorrect.
- [x] Needs shipping - Just check if "Shipping Address" section is not displayed to the user if product doesn't require Shipping like virtual products.
- [x] Virtual with Sale price
- [x] Sold individually
- [x] Virtual sold individually
- [x] Without enough stock
- [x] With invalid shipping address
- [x] Regular Price with tax
- [x] Regular Price with tax and quantity
- [x] With coupon (fixed rate and percentage)
- [x] PRB: Default, Branded, etc
- [x] Descriptor
- [x] Cart page
- [x] With Product Add-Ons
  - Note: It didn't work before. it only worked when changing the quantity. No need to do that anymore it works just by picking an add-on now.
- [x] Checkout block/Express checkout

#### Keep in mind

- [x] https://github.com/woocommerce/woocommerce-gateway-stripe/pull/727
- [x] https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1100

#### Observations

- `wc_stripe_get_selected_product_data` is not returning the same response as `wc_stripe_add_to_cart`

---
